src/Extension.Tests/CodeRunEvaluationTests.cs

```diff
@@ -0,0 +1,290 @@
+{
+    (0, 0, 0),
+    (0.5, 1.0, 0.25),
+    (100, 50, 2500)
+};
+
+testcases.All(testcase => triangleArea(testcase.Item1, testcase.Item2) == testcase.Item3)
+");
+
+            // students code runs
+            await kernel.SubmitCodeAsync(@"
+#!question 5
+
+double triangleArea(double x, double y) 
+{
+    return x * y;
+}
+");
+
+            events.Should().ContainSingle<CodeRunEvaluationProduced>()
+                .Which.Evaluation.Passed.Should().Be(false);
+        }
+
+        [Fact]
+        public async Task submitting_code_that_passes_evaluation_criterion_gives_passed_evaluation()
+        {
+            var evaluator = new Evaluator();
+            using var csharpKernel = new CSharpKernel();
+            using var kernel = new CompositeKernel { csharpKernel };
+            kernel
+                .UseQuestionMagicCommand(evaluator)
+                .UseAnswerMagicCommand(evaluator);
+
+
+            using var events = kernel.KernelEvents.ToSubscribedList();
+
+            // instructor provides testcase script which returns bool
+            await kernel.SubmitCodeAsync(@"
+#!answer 6
+var testcases = new (double, double, double)[]
+{
+    (0, 0, 0),
+    (0.5, 1.0, 0.25),
+    (100, 50, 2500)
+};
+
+testcases.All(testcase => triangleArea(testcase.Item1, testcase.Item2) == testcase.Item3)
+");
+
+            // students code runs
+            await kernel.SubmitCodeAsync(@"
+#!question 6
+
+double triangleArea(double x, double y) 
+{
+    return 0.5 * x * y;
+}
+");
+
+            events.Should().ContainSingle<CodeRunEvaluationProduced>()
+                .Which.Evaluation.Passed.Should().Be(true);
+        }
+
+
+        [Fact]
+        public async Task forward_symbol_declaration_will_not_affect_evaluation_criteria_runs()
+        {
+            var evaluator = new Evaluator();
+            using var csharpKernel = new CSharpKernel();
+            using var kernel = new CompositeKernel { csharpKernel };
+            kernel
+                .UseQuestionMagicCommand(evaluator)
+                .UseAnswerMagicCommand(evaluator);
+
+            using var events = kernel.KernelEvents.ToSubscribedList();
+
+            await kernel.SubmitCodeAsync(@"
+double triangleArea(double x, double y) => 0.0
+");
+
+            await kernel.SubmitCodeAsync(@"
+#!answer 6
+var testcases = new (double, double, double)[]
+{
+    (0, 0, 0),
+    (0.5, 1.0, 0.25),
+    (100, 50, 2500)
+};
+
+testcases.All(testcase => triangleArea(testcase.Item1, testcase.Item2) == testcase.Item3)
+");
+
+            // students code runs
+            await kernel.SubmitCodeAsync(@"
+#!question 6
+
+double triangleArea(double x, double y) 
+{
+    return 0.5 * x * y;
+}
+");
+
+            events.Should().ContainSingle<CodeRunEvaluationProduced>()
+                .Which.Evaluation.Passed.Should().Be(true);
+        }
+
+        [Fact]
+        public async Task submitting_code_without_setting_evaluation_criterion_gives_passed_evaluation()
```

Comment:
Why should this pass? Does this not indicate a bug on the part of the evaluation code, i.e. missing evaluation?

If it is a bug, who is the audience for the error that should result? How do we make sure it surfaces at the right time?

Reply:
Evaluation is student-facing so it seems reasonable to give passing if no criterion is set. I agree it is important for professors to discover questions with no criterion set. This might be worth a discussion. The current model is we assume the main (student-facing) notebook is run after the supporting notebook.

Nerdy sidenote: the current invariant is "evaluation is passing if and only if all criteria is true." The statement "all criteria is true" is vacuously true if there are no criteria.

Reply:
re: the nerdy side node, a question with no criteria sounds like an invalid state. How do we prevent the teacher notebook from allowing this state?

Reply:
Yeah, I agree. and in general, the process of how a teacher can verify that the notebooks he built runs correctly is a serious consideration.